vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go

```diff
@@ -0,0 +1,276 @@
+
+	"k8s.io/kube-openapi/pkg/util"
+)
+
+const (
+	definitionPrefix = "#/definitions/"
+)
+
+// Run a walkRefCallback method on all references of an OpenAPI spec
+type referenceWalker struct {
+	// walkRefCallback will be called on each reference and the return value
+	// will replace that reference. This will allow the callers to change
+	// all/some references of an spec (e.g. useful in renaming definitions).
+	walkRefCallback func(ref spec.Ref) spec.Ref
+
+	// The spec to walk through.
+	root *spec.Swagger
+}
+
+func walkOnAllReferences(walkRef func(ref spec.Ref) spec.Ref, sp *spec.Swagger) {
+	walker := &referenceWalker{walkRefCallback: walkRef, root: sp}
+	walker.Start()
+}
+
+func (s *referenceWalker) walkRef(ref spec.Ref) spec.Ref {
+	if ref.String() != "" {
+		refStr := ref.String()
+		// References that start with #/definitions/ has a definition
+		// inside the same spec file. If that is the case, walk through
+		// those definitions too.
+		// We do not support external references yet.
+		if strings.HasPrefix(refStr, definitionPrefix) {
+			def := s.root.Definitions[refStr[len(definitionPrefix):]]
+			s.walkSchema(&def)
+		}
+	}
+	return s.walkRefCallback(ref)
+}
+
+func (s *referenceWalker) walkSchema(schema *spec.Schema) {
+	if schema == nil {
+		return
+	}
+	schema.Ref = s.walkRef(schema.Ref)
+	for _, v := range schema.Definitions {
+		s.walkSchema(&v)
+	}
+	for _, v := range schema.Properties {
+		s.walkSchema(&v)
+	}
+	for _, v := range schema.PatternProperties {
+		s.walkSchema(&v)
+	}
+	for _, v := range schema.AllOf {
+		s.walkSchema(&v)
+	}
+	for _, v := range schema.AnyOf {
+		s.walkSchema(&v)
+	}
+	for _, v := range schema.OneOf {
+		s.walkSchema(&v)
+	}
+	if schema.Not != nil {
+		s.walkSchema(schema.Not)
+	}
+	if schema.AdditionalProperties != nil && schema.AdditionalProperties.Schema != nil {
+		s.walkSchema(schema.AdditionalProperties.Schema)
+	}
+	if schema.AdditionalItems != nil && schema.AdditionalItems.Schema != nil {
+		s.walkSchema(schema.AdditionalItems.Schema)
+	}
+	if schema.Items != nil {
+		if schema.Items.Schema != nil {
+			s.walkSchema(schema.Items.Schema)
+		}
+		for _, v := range schema.Items.Schemas {
+			s.walkSchema(&v)
+		}
+	}
+}
+
+func (s *referenceWalker) walkParams(params []spec.Parameter) {
+	if params == nil {
+		return
+	}
+	for _, param := range params {
+		param.Ref = s.walkRef(param.Ref)
+		s.walkSchema(param.Schema)
+		if param.Items != nil {
+			param.Items.Ref = s.walkRef(param.Items.Ref)
+		}
+	}
+}
+
+func (s *referenceWalker) walkResponse(resp *spec.Response) {
+	if resp == nil {
+		return
+	}
+	resp.Ref = s.walkRef(resp.Ref)
+	s.walkSchema(resp.Schema)
+}
+
+func (s *referenceWalker) walkOperation(op *spec.Operation) {
+	if op == nil {
+		return
+	}
+	s.walkParams(op.Parameters)
+	if op.Responses == nil {
+		return
+	}
+	s.walkResponse(op.Responses.Default)
+	for _, r := range op.Responses.StatusCodeResponses {
+		s.walkResponse(&r)
+	}
+}
+
+func (s *referenceWalker) Start() {
```

Comment:
Do you want that test to be added before this PR merged? I am sending another PR to kube-openapi repo to address your comments, but I don't think we need to keep this PR blocked on that.